general/mem_manager.hpp

```diff
@@ -13,33 +13,462 @@
+    * When destroyed, a Memory object does NOT automatically delete any
+      allocated memory.
+    * Only the method Delete() will deallocate a Memory object.
+    * Other methods that modify the object (e.g. New(), Wrap(), etc) will simply
+      overwrite the old contents.
+    * One difference with a pointer is that a const Memory object is NOT like a
+      const pointer: it cannot be used to overwrite entries in the memory.
+
+    A Memory object stores up to two different pointers: one host pointer (with
+    MemoryType from MemoryClass::HOST) and one device pointer (currently one of
+    MemoryType::CUDA or MemoryTyep::CUDA_UVM).
+
+    A Memory object can hold (wrap) an externally allocated pointer with any
+    given MemoryType.
+
+    Access to the content of the Memory object can be requested with any given
+    MemoryClass through the methods ReadWriteAccess(), ReadAccess(), and
+    WriteAccess(). Requesting such access may result in additional (internally
+    handled) memory allocation and/or memory copy.
+    * When ReadWriteAccess() is called, the returned pointer becomes the only
+      valid pointer.
+    * When ReadAccess() is called, the returned pointer becomes valid, however
+      the other pointer (host or device) may remain valid as well.
+    * When WriteAccess() is called, the returned pointer becomes the only valid
+      pointer, however, unlike ReadWriteAccess(), no memory copy will be
+      performed.
+
+    The host memory (pointer from MemoryClass::HOST) can be accessed through the
+    inline methods: `operator[]()`, `operator*()`, the implicit conversion
+    functions `operator T*()`, `operator const T*()`, and the explicit
+    conversion template functions `operator U*()`,  `operator const U*()` (with
+    any suitable type U). In certain cases, using these methods may have
+    undefined behavior, e.g. if the host pointer is not currently valid. */
+template <typename T>
+class Memory
+{
+protected:
+   friend class MemoryManager;
+   friend void MemoryPrintFlags(unsigned flags);
+
+   enum FlagMask
+   {
+      REGISTERED    = 1,   ///< #h_ptr is registered with the MemoryManager
+      OWNS_HOST     = 2,   ///< The host pointer will be deleted by Delete()
+      OWNS_DEVICE   = 4,   ///< The device pointer will be deleted by Delete()
+      OWNS_INTERNAL = 8,   ///< Ownership flag for internal Memory data
+      VALID_HOST    = 16,  ///< Host pointer is valid
+      VALID_DEVICE  = 32,  ///< Device pointer is valid
+      ALIAS         = 64,
+      EXEC_FLAG     = 128  ///< Optional execution flag used by MFEM
+   };
+
+   /// Pointer to host memory. Not owned.
+   /** When the pointer is not registered with the MemoryManager, this pointer
+       has type MemoryType::HOST. When the pointer is registered, it can be any
+       type from MemoryClass::HOST. */
+   T *h_ptr;
+   int capacity;
+   mutable unsigned flags;
+   // 'flags' is mutable so that it can be modified in Set{Host,Device}PtrOwner,
```

Comment:
It is necessary, if you want to avoid using the `mfem::Device` vector kernels for all `Vectors`, and we definitely do not want to launch kernels for small vectors.

Reply:
Why not just checking the size of the vectors then? If small vectors are the only reasons for all this I think we can find much simpler alternatives.